File: js/src/maplestory/api/msea/dto/character/characterSetEffect.ts
import * as base from '../../../common/dto/character/characterSetEffect';
import {
  CharacterSetEffectBody,
  CharacterSetEffectInfoBody,
  CharacterSetEffectOptionFullBody,
  CharacterSetEffectSetBody,
} from '../../response/character/characterSetEffectBody';

/**
 * Character information about equipped set item effects
 */
export class CharacterSetEffectDto extends base.CharacterSetEffectDto {
  /**
   * Reference date for query (SGT, daily data with hours and minutes set to 0)
   */
  public override date: Date | null;

  /**
   * Set Item effect information
   */
  public override setEffect: CharacterSetEffectSetDto[];

  constructor(obj: CharacterSetEffectBody) {
    super();

    const { date, set_effect } = obj;
    this.date = date ? new Date(date) : null;
    this.setEffect = set_effect.map(
      (effect) => new CharacterSetEffectSetDto(effect),
    );
  }
}

/**
 * Active Set Item effect information
 */
export class CharacterSetEffectInfoDto extends base.CharacterSetEffectInfoDto {
  /**
   * Set Item effect level (number of equipped items)
   */
  public override setCount: number;

  /**
   * Set Item effect
   */
  public override setOption: string;

  constructor(obj: CharacterSetEffectInfoBody) {
    super();

    const { set_count, set_option } = obj;

    this.setCount = set_count;
    this.setOption = set_option;
  }
}

/**
 * All Set Item effect information
 */
export class CharacterSetEffectOptionFullDto extends base.CharacterSetEffectOptionFullDto {
  /**
   * Set Item effect level (number of equipped items)
   */
  public override setCount: number;

  /**
   * Set Item effect
   */
  public override setOption: string;

  constructor(obj: CharacterSetEffectOptionFullBody) {
    super();

    const { set_count, set_option } = obj;

    this.setCount = set_count;
    this.setOption = set_option;
  }
}

/**
 * Set Item effect information
 */
export class CharacterSetEffectSetDto extends base.CharacterSetEffectSetDto {
  /**
   * Set Item effect name
   */
  public override setName: string;

  /**
   * Number of Set Items (including Lucky Items)
   */
  public override totalSetCount: number;

  /**
   * Active Set Item effect information
   */
  public override setEffectInfo: CharacterSetEffectInfoDto[];

  /**
   * All Set Item effect information
   */
  public override setOptionFull: CharacterSetEffectOptionFullDto[];

  constructor(obj: CharacterSetEffectSetBody) {
    super();

    const { set_name, total_set_count, set_effect_info, set_option_full } = obj;

    this.setName = set_name;
    this.totalSetCount = total_set_count;
    this.setEffectInfo = set_effect_info.map(
      (info) => new CharacterSetEffectInfoDto(info),
    );
    this.setOptionFull = set_option_full.map(
      (info) => new CharacterSetEffectOptionFullDto(info),
    );
  }
}

File: csharp/MapleStory.OpenAPI/Src/MSEA/Dto/Character/CharacterSetEffect.cs
using Newtonsoft.Json;
using Base = MapleStory.OpenAPI.Common.DTO;

namespace MapleStory.OpenAPI.MSEA.DTO
{
    /// <summary>
    /// Character information about equipped set item effects
    /// </summary>
    public class CharacterSetEffectDTO : Base.CharacterSetEffectDTO<CharacterSetEffectSetDTO>
    {
        /// <summary>
        /// Reference date for query (SGT, daily data with hours and minutes set to 0)
        /// </summary>
        [JsonProperty("date")]
        public override DateTimeOffset? Date
        {
            get
            {
                return _date?.ToOffset(TimeSpan.FromHours(8));
            }
            set
            {
                _date = value;
            }
        }

        private DateTimeOffset? _date;

        /// <summary>
        /// Set Item effect information
        /// </summary>
        [JsonProperty("set_effect")]
        [JsonConverter(typeof(NullAsEmptyConverter))]
        public override List<CharacterSetEffectSetDTO> SetEffect { get; set; } = new();
    }

    /// <summary>
    /// Active Set Item effect information
    /// </summary>
    public class CharacterSetEffectInfoDTO : Base.CharacterSetEffectInfoDTO
    {
        /// <summary>
        /// Set Item effect level (number of equipped items)
        /// </summary>
        [JsonProperty("set_count")]
        public override int SetCount { get; set; }

        /// <summary>
        /// Set Item effect
        /// </summary>
        [JsonProperty("set_option")]
        public override string SetOption { get; set; }
    }

    /// <summary>
    /// All Set Item effect information
    /// </summary>
    public class CharacterSetEffectOptionFullDTO : Base.CharacterSetEffectOptionFullDTO
    {
        /// <summary>
        /// Set Item effect level (number of equipped items)
        /// </summary>
        [JsonProperty("set_count")]
        public override int SetCount { get; set; }

        /// <summary>
        /// Set Item effect
        /// </summary>
        [JsonProperty("set_option")]
        public override string SetOption { get; set; }
    }

    /// <summary>
    /// Set Item effect information
    /// </summary>
    public class CharacterSetEffectSetDTO : Base.CharacterSetEffectSetDTO<CharacterSetEffectInfoDTO, CharacterSetEffectOptionFullDTO>
    {
        /// <summary>
        /// Set Item effect name
        /// </summary>
        [JsonProperty("set_name")]
        public override string SetName { get; set; }

        /// <summary>
        /// Number of Set Items (including Lucky Items)
        /// </summary>
        [JsonProperty("total_set_count")]
        public override long TotalSetCount { get; set; }

        /// <summary>
        /// Active Set Item effect information
        /// </summary>
        [JsonProperty("set_effect_info")]
        [JsonConverter(typeof(NullAsEmptyConverter))]
        public override List<CharacterSetEffectInfoDTO> SetEffectInfo { get; set; } = new();

        /// <summary>
        /// All Set Item effect information
        /// </summary>
        [JsonProperty("set_option_full")]
        [JsonConverter(typeof(NullAsEmptyConverter))]
        public override List<CharacterSetEffectOptionFullDTO> SetOptionFull { get; set; } = new();
    }
}
